### ISSUE TITLE
parameters for "ammo_restriction" should have a default when a parameter isn't given

### ISSUE THREAD

**Issue by catdach** [CleverRaven/Cataclysm-DDA#53375]
### Describe the bug

All guns using an integral magazine are required to have a pocket with a defined `"ammo_restriction": { "ammo_type": count }` despite the fact that both `"ammo_type"` and `count` are defined within gun itself as `"ammo"` and `"clip_size"` respectively.

### Steps to reproduce

1. remove the `"ammo_restriction": { "ammo_type": count }` from any  shotgun's pocket data
2. start a new game with any scenario that doesn't start you with nothing, 
3. spawn in that gun and some extra ammo
4. unload the gun
5. try to reload the gun 
6. for sake of the experiment, ignore all error messages 
7. if the game hasn't crashed, note that the gun will take pretty much anything as ammo in any amount

### Expected behavior

`"ammo_type"` and `count` should default to the values defined in `"ammo"` and `"clip_size"` **when and only when**, they aren't explicitly defined by `"ammo_restriction"`

### Versions and configuration

- OS: Windows
    - OS Version: 10.0 2009
- Game Version: 61fa86a [64-bit]
- Graphics Version: Tiles
- Game Language: System language []
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    No Fungal Growth [no_fungal_growth],
    Bionic Professions [package_bionic_professions],
    Aftershock [aftershock],
    Aftershock: Exoplanet [aftershock_exoplanet]
]

### Additional context

In most cases it doesn't really matter, but it's really weird and arbitrary to define these values manually. It could make sense if there were guns with multiple integrated magazine pockets that each took different ammo or different amounts of ammo, but even then there isn't a reason to not have any default. At this point `"ammo"` and `"clip_size"` seem to do effectively nothing. 

This change would be important for gun mods that will need to change other parts pocket data (and yes there are other parts that are meaningful besides the ammo restriction) via `"pocket_mods"` (which overwrites **all** the pocket data), but doesn't want to change parts of the pocket data that would vary depending the specific gun that the mod is attached to (AKA the shotgun speedloader chute).

Edit: I've been trying to fix shotgun speedloaders for weeks, and this is pretty necessary in order to do it.

**Comment by Photoloss** [comment]
Are the two field sets actually the same? Think of something like the larger [Super Soakers](http://www.isoaker.com/Armoury/Analysis/1996/super_soaker_xp105.php). It has both a large water tank serving as a reservoir which can hold a large amount of fluid but 1 l of water is a "count" of 4 units while 1 l of ethanol is a "count" of 1000 or so units. Then it also has the smaller pressure chamber which is the part you actually make ready to shoot via pumping. In this case both can potentially be filled to increase the total ammo capacity which might not be safe enough in actual weapons, but either way it is a clear example of a weapon with completely integrated ammo storage and different values for "clip size" and "ammo capacity".

If the existing fields *are* supposed to be the same then perhaps it might be more useful to implement the above functionality, also adding the option of independent "ammo per shot" costs to each pocket entry. If nothing else this could still be used in vanilla C:DDA to represent guns which can hold an extra round in the chamber.

**Comment by catdach** [comment]
> Are the two field sets actually the same? Think of something like the larger [Super Soakers](http://www.isoaker.com/Armoury/Analysis/1996/super_soaker_xp105.php). It has both a large water tank serving as a reservoir which can hold a large amount of fluid but 1 l of water is a "count" of 4 units while 1 l of ethanol is a "count" of 1000 or so units. Then it also has the smaller pressure chamber which is the part you actually make ready to shoot via pumping. In this case both can potentially be filled to increase the total ammo capacity which might not be safe enough in actual weapons, but either way it is a clear example of a weapon with completely integrated ammo storage and different values for "clip size" and "ammo capacity".
> 
> If the existing fields _are_ supposed to be the same then perhaps it might be more useful to implement the above functionality, also adding the option of independent "ammo per shot" costs to each pocket entry. If nothing else this could still be used in vanilla C:DDA to represent guns which can hold an extra round in the chamber.

This sounds is similar to the case I addressed in the "additional context" (Having multiple integrated magazine pockets, Which I tried to say was a case where defining `"ammo_restiction"` does indeed make sense) with the caveat of liquid ammunition which can differ in units per Liter ratio. There a couple ways to address this;
1. The closest thing in the game to a Water gun that I'm aware of is the Chemical Thrower and Flame Thrower, Which have their own ammo type named `"chemical_spray"` and `"flammable"` respectively, and that makes it a standardized `count` (also neither use an integral magazine, so it's not what I'm addressing anyway)
2. You could restrict the magazine contents by Volume instead of count like other liquid containers already do (cans, bottles, etc.).

To be clear, I think pocket data defining stuff is fine and makes sense most of the time. The problem is that right now it is absolutely mandatory, so the "ammo" and "clip_size" fields do absolutely nothing.

**Comment by Photoloss** [comment]
> This sounds is similar to the case I addressed in the "additional context" (Having multiple integrated magazine pockets

Not quite the same as the two tanks are in sequence rather than in parallel. The reservoir only exists to transfer ammo to the pressure tank, the smaller tank essentially acts as the actual chamber of the gun including the risks of being "primed". Only firearms don't really support chambering multiple rounds, I suppose the closest equivalent would be a double-barrelled shotgun with a larger manual-action magazine feeding both barrels, you get two shots off a reload but the gun as a whole could hold a dozen rounds.

> I think we call that a firing mode? Ever use an under barrel shotgun/grenade launcher?

Can firing modes do that? I was mainly thinking of a built-in means of doing away with the external standardisation you described for flamethrowers. In the same vein ammo pouches limit the number of rounds per ammo type, and in contrast to underbarrel attachments the pouch still becomes completely full if you exhaust its capacity for a single type of ammo. Except the pouch does not actually *fire* the ammo of course.

> I'm not sure that you understand what I'm asking for. To be clear, I think pocket data defining stuff is fine and makes sense most of the time. The problem is that right now it is absolutely mandatory, so the "ammo" and "clip_size" fields do absolutely nothing.

I understand you just want it removed because it seems to serve no purpose. I am arguing that it *can* serve a specific purpose, whether or not it actually works that way in game right now.

**Comment by catdach** [comment]
>>"ammo_type" and count should default to the values defined in "ammo" and "clip_size" **when they aren't explicitly defined** by "ammo_restriction"
>> To be clear, I think **pocket data defining stuff is fine** and makes sense most of the time.
> I understand **you just want it removed** because it seems to serve no purpose.
???

**Comment by catdach** [comment]
I don't want it removed, I just don't want alternatives.

**Comment by Photoloss** [comment]
> I don't want it removed, I just want alternatives.

Okay I should have been more precise: you want the redundant requirements removed.

In practice I can't think of a vanilla use case where you would be able to use the gun-level parameters without having a magazine pocket, notably even bows and crossbows have single-shot pockets. So to me the simplest solution seems to be to remove the gun parameters entirely and require the ones within the magazine instead. A counterexample would be a UPS-powered gun with the ability to pre-charge multiple shots into an internal battery, but to my knowledge none of those exist in game and the proper implementation would be to have multiple magazine pockets as you describe under "Additional context" and allow UPS feeding of said pockets.

**Comment by catdach** [comment]
> you want the redundant requirements removed.

No, I want the redundant requirements **_TO WORK_**

**Comment by Photoloss** [comment]
Just for funzies I set the ammo types to conflicting values. Throws a bunch of errors despite both types being valid so the redundant requirement definitely does "something". The gun settings show up in the UI while the pocket initialisation fails completely. I can only assume they serve distinct purposes in the underlying code because throwing errors like this is a *terrible* consistency check.

**Comment by catdach** [comment]
ok, I actually didn't expect that. so ya they don't do nothing, I just want them to also do their job and not just throw errors when "ammo_restriction" isn't defined.

**Comment by catdach** [comment]
> In practice I can't think of a vanilla use case where you would be able to use the gun-level parameters without having a magazine pocket

I'll tell you a secret, I made this issue request mainly so that I can get shotgun speedloaders to work properly. The fact that `"ammo_restriction": { "ammo_type": count }` of the pocket_data takes priority over the `"clip_size"` of the gun, is the one thing that's keeping me from fixing them entirely.

**Comment by Photoloss** [comment]
> I'll tell you a secret, I made this issue request mainly so that I can get shotgun speedloaders to work properly.

That sounds a bit like my "magazine double-barrelled" example. You basically want to load 2 shells at the same time, ready to fire, but a magazine in the shotgun might hold more than 2 which take longer to load into the barrels, correct?

> The problem is that gunmods currently cannot access the magazine pocket's parameters in a meaningful way, without rewriting the whole pocket magazine from scratch, within the gunmod's own data. Which is fine if all the magazine pocket's parameters always remain consistent for every gun that the gunmod can be attached to and nothing else would every change them, including other gunmods. Enter the shotgun speedloader chute. It can be attached to a ton of shotguns with different maximum ammo. Hence, my issue request.

That is a separate issue. The root of your problem is not that the gun parameters don't work well, but rather that gunmods do not have a way to "extend" pocket objects the same way they work on weapon or ammo flags. I imagine contributors would encounter the same issue when attempting to implement a "one in the chamber" mod which boosts magazine size by 1 at the cost of reliability, or a "high pressure tank" mod for multiple gas-ammo weapons with integrated tanks of different sizes.

**Comment by catdach** [comment]
So let me elaborate my fix is basically to add this line to the json file for the chute item
```
...
    "pocket_mods": [ 
 {
 "pocket_type": "MAGAZINE",
 "acceptable_ammo": [ "shot" : 8 ],
 "allowed_speedloaders": [ "shot_speedloader6", "shot_speedloader8" ] 
 } 
],
...
```
it seems to work really well aside from the fact that it forces the max ammo to be 8. feel free to try it for yourself.

>Or just instantly dump say 6 shells into a shotgun which holds 4 in its internal magazine, leaving 2 in the speedloader?

So this, surprisingly, is exactly what happens if `"acceptable_ammo":` is set to `[ "shot" : 4 ],` So that already works fine.

>That is a separate issue. The root of your problem is not that the gun parameters don't work well, but rather that gunmods do not have a way to "extend" pocket objects the same way they work on weapon or ammo flags.

Yes, that is also an issue, but if what I'm requesting is implemented, Then I wouldn't need a way to "extend" pocket objects in this case. Out of the two, this issue seemed like it would be an easier and more reasonable thing to implement. Which in hindsight may have been wrong but I'm not exactly thinking clearly after trying to figure out how to do it myself for the past few weeks.

There was an old PR that sort of claimed to allow the  `"pocket_mods"` to "_modify_" pockets, but for magazine type pockets it just replaces them from what I can tell.

**Comment by Hirmuolio** [comment]
I think `"clip_size"` is completely useless attribute and should be removed. The ammo restriction in the pocket has the same data and is better place for that data. I am pretty sure `"clip_size"` is never used during reloading.

`"ammo"` is almost useless attribute. The only place where "normal" guns need it is when reloading with magazines. The magazine well pocket only knows what magazine is compatible, it does not know what ammo the magazine should have to contain (You wouldn't want to reload AR-15 with STANAG amgazine fiiled with 9mm ammo). Not sure if that is problem for any magazine in game.

I think the best solution would be to just get rid of both `"clip_size"` and `"ammo"`.
Add some sort of `valid_ammo` property to magazine wells. This way magazine reloading would have all the data it needs from the pocket alone.

But removing `"ammo"` is not exactly easy task becaus not all guns are "normal".
Take a look at this thing and check what it uses as ammo: https://github.com/CleverRaven/Cataclysm-DDA/blob/master/data/json/items/gun/shot.json#L40
It has `"ammo": [ "shot" ]` (from shotgun base) and additional `"ups_charges": 1`.
It consumes TWO different things for every shot: One shotgun shell and one charge of UPS.

Currently there is no definition for what ammo gun uses to fire. When you fire a gun it just uses any ammo contained in it. And the way these UPS+ammo combos work is a bit hacky (and does involve `"ammo"` attribute. It doesn't care what the ammo is, it just checks that the attribute exists).

For proper fix I think we need proper support for multi ammo and multi magazine guns.

Fun detail: `"ammo"` on guns and `"ammo"` on tools are two completely separate things that have nothing in common.

**Comment by Photoloss** [comment]
I still think "clip size" *should* be a usable concept, even if the code is migrated to pockets instead. How else do you represent a double-barrelled shotgun with an 8 round magazine feeding both barrels?

**Comment by Hirmuolio** [comment]
One gun with two barrels each with their own magazines?

Gun with one magazine, integrated gunmod with second magazine.  
Gun normal fire mode fires itself, gunmod normal firemode fires itself.  
Gun special firemode fires itself and gunmod simultaneously.

**Comment by Photoloss** [comment]
No, one magazine feeding both barrels, with manual pump/bolt action so you can't just make it semi-auto with a 1- and 2-shot fire mode. So you get two shots "instantly", reload those two in sequence with a quick action, and can fire a total of 8 (10 if you keep rounds in the chambers) shots before needing to swap magazines.

This might also see use with internal multi-shot capacitors, heatsinks, pressure vessels and other non-firearm ranged weapons. Especially with UPS-powered lasers since they have no internal magazine and draw electricity instantly when fired.

**Comment by catdach** [comment]
@Photoloss I'm sorry, but what you're saying doesn't make any sense. 
Right now there is no meaningful distinction in-game between the camber of a gun and its magazine, especially for guns with integral magazines like shotguns. 
Neither is there a meaningful in-game distinction between manual and semiautomatic actions of guns. 
Neither is there a meaningful in-game distinction between a semi-auto 2-shot fire mode and an "instant" 2 shots from a double barreled shotgun.
What you have described is functionally the same as having any current shotgun with, a maximum ammo capacity of 10, and a 2-shot fire mode. 

@Hirmuolio At this point I personally don't care that much about the state of `"ammo"` or `"clip_size"` anymore. They have become the focus of discussion in a way that I did not intend. My focus is `"ammo_restriction": { "ammo_type": count }` which effectively renders `"ammo"` or `"clip_size"` obsolete in _most_ circumstances. In wanting for ``"ammo_restriction":` to have default parameters, `"ammo"` and `"clip_size"` seemed like the obvious answer, and was surprised when I found that they weren't fulfilling this role already. I figured that I should mention them just in case someone was going to remove them for their redundancy. Giving them a use seems better than removing them. that being said, if there is a better solution then I don't care if they're gotten rid of. I would say that there is use to having such parameters tied to the gun itself rather than its pocket, but I can't think of a tangible example if my issue is otherwise resolved.

> "ammo" is almost useless attribute. The only place where "normal" guns need it is when reloading with magazines. The magazine well pocket only knows what magazine is compatible, it does not know what ammo the magazine should have to contain (You wouldn't want to reload AR-15 with STANAG magazine filed with 9mm ammo). Not sure if that is problem for any magazine in game.

Maybe the blackout 300 conversion kits make use of this? Perhaps magazine adaptors? As far as I know, gunmods cannot access or modify any pocket data, only data in the overall gun.
